src/librustc/ty/subst.rs

```diff
@@ -81,80 +96,81 @@ impl<'tcx> Kind<'tcx> {
     }
 
     #[inline]
-    pub fn as_type(self) -> Option<Ty<'tcx>> {
-        unsafe {
-            self.downcast(TYPE_TAG)
-        }
-    }
-
-    #[inline]
-    pub fn as_region(self) -> Option<ty::Region<'tcx>> {
-        unsafe {
-            self.downcast(REGION_TAG)
+    pub fn unpack(self) -> Option<UnpackedKind<'tcx>> {
```

Comment:
The existence of this method should imply two things: `downcast` isn't separately needed *and* it should always succeed (`bug!` or `intrinsics::unreachable()` - since it should be impossible to craft a bad `Kind`).